### PR TITLE
Switch WorkPool of ConsumerWorkService to channels

### PR DESCRIPTION
**Pull request by danielmarbach**
## Proposed Changes

Switches WorkPool of ConsumerWorkService to channels

As discussed with @stebet in conversation over here https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/866#discussion_r444092965

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

**Comment by stebet** [comment]
Looks good to me :)

File: projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs [hunk 1]
```diff
@@ -126,36 +113,37 @@ async Task Loop()
...
+                    // ignored
+                }
             }
 
             static async Task OffloadToWorkerThreadPool(Action action, SemaphoreSlim limiter)
             {
                 try
                 {
-                    await Task.Run(() => action());
+                    // like Task.Run but doesn't closure allocate
```
Comment (bollhals) [inline projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs:142]: 🎉  I like =)

**Comment by lukebakken** [comment]
Thanks everyone!

**Comment by lukebakken** [comment]
Backported to `6.x` branch.

File: projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs (hunk 1 above)
Comment (sungam3r) [reply projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs:142]: Unfortunately, it still allocates delegate. IIRC it is well known issue of Roslyn. If you really do not want allocate delegate each call then define static delegate and use it here.

Comment (michaelklishin) [reply projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs:142]: @sungam3r thanks for chiming in. You are welcome to submit a PR that changes the comment (and links to the Roslyn issue) or even switched to use a static delegate if it's a safe thing to do in this context. Thank you.

Comment (danielmarbach) [reply projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs:142]: Comparison according to sharplab

```
    [Serializable]
    [CompilerGenerated]
    private sealed class <>c
    {
        public static readonly <>c <>9 = new <>c();

        public static Action<object> <>9__0_0;

        internal void <TaskFactory>b__0_0(object state)
        {
            ((Action)state)();
        }
    }

awaiter = Task.Factory.StartNew(<>c.<>9__0_0 ?? (<>c.<>9__0_0 = new Action<object>(<>c.<>9.<TaskFactory>b__0_0)), action, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default).GetAwaiter();
```

so the `Action` allocation happens once and is then cached for further execution in a static Action field.

vs `Task.Run(() => action())`

```
                    <>c__DisplayClass1_0 <>c__DisplayClass1_ = new <>c__DisplayClass1_0();
                    <>c__DisplayClass1_.action = action;
                    awaiter = Task.Run(new Action(<>c__DisplayClass1_.<TaskRun>b__0)).GetAwaiter();
```

display class creation for every run plus action allocation for every run

Comment (sungam3r) [reply projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs:142]: I was wrong. Such lambda usage is OK. I re-read the Roslyn issue and remembered details.

**Pull request merged**